File: python/with_java/programmers/level 3/정수 삼각형.py
def solution(triangle):
    dp = []
    for row in triangle:
        if not dp:
            dp = row
            continue
        temp = []
        for i in range(len(row)):
            if i == 0:
                temp.append(dp[i] + row[i])
            elif i == len(row)-1:
                temp.append(dp[i-1] + row[i])
            else:
                temp.append(max(dp[i-1], dp[i]) + row[i])
        dp = temp
    return max(dp)

File: java/src/main/java/ps_traning/programmers/level_3/정수_삼각형.java
package ps_traning.programmers.level_3;

import java.util.Arrays;

public class 정수_삼각형 {
    public int solution(int[][] triangle) {
        int answer = 0;
        int[] dp = {0};
        for (int[] row : triangle) {
            int[] temp = new int[row.length];

            for (int i = 0; i < row.length; i++) {
                if (i == 0) {
                    temp[i] = dp[i] + row[i];
                } else if (i == row.length - 1) {
                    temp[i] = dp[i - 1] + row[i];
                } else {
                    temp[i] = Math.max(dp[i], dp[i - 1]) + row[i];
                }
            }

            dp = temp;
        }
        Arrays.sort(dp);
        answer = dp[dp.length - 1];
        return answer;
    }
}
